--- dhcp-4.4.2/client/dhclient.c.extravars	2021-04-02 12:50:00.989139908 +0000
+++ dhcp-4.4.2/client/dhclient.c		2021-04-02 12:55:15.477409996 +0000
@@ -89,8 +89,6 @@ static const char message [] = "Internet
 static const char url [] = "For info, please visit https://www.isc.org/software/dhcp/";
 #endif /* UNIT_TEST */
 
-u_int16_t local_port = 0;
-u_int16_t remote_port = 0;
 #if defined(DHCPv6) && defined(DHCP4o6)
 int dhcp4o6_state = -1; /* -1 = stopped, 0 = polling, 1 = started */
 #endif
--- dhcp-4.4.2/relay/dhcrelay.c.extravars	2021-04-02 13:05:13.513727484 +0000
+++ dhcp-4.4.2/relay/dhcrelay.c			2021-04-02 13:05:33.690494049 +0000
@@ -95,9 +95,6 @@ enum { forward_and_append,	/* Forward an
        forward_untouched,	/* Forward without changes. */
        discard } agent_relay_mode = forward_and_replace;
 
-u_int16_t local_port;
-u_int16_t remote_port;
-
 /* Relay agent server list. */
 struct server_list {
 	struct server_list *next;
--- dhcp-4.4.2/server/mdb.c.extravars	2019-12-17 19:13:31.000000000 +0000
+++ dhcp-4.4.2/server/mdb.c		2021-04-02 13:09:26.508311259 +0000
@@ -67,8 +67,6 @@ static host_id_info_t *host_id_info = NU
 
 int numclasseswritten;
 
-omapi_object_type_t *dhcp_type_host;
-
 isc_result_t enter_class(cd, dynamicp, commit)
 	struct class *cd;
 	int dynamicp;
