diff -urNp -x '*.orig' dhcp-4.3.5.org/common/lpf.c dhcp-4.3.5/common/lpf.c
--- dhcp-4.3.5.org/common/lpf.c	2016-09-27 21:16:50.000000000 +0200
+++ dhcp-4.3.5/common/lpf.c	2021-04-01 22:09:12.139265409 +0200
@@ -247,6 +247,7 @@ static void lpf_gen_filter_setup (info)
 	struct interface_info *info;
 {
 	struct sock_fprog p;
+	memset(&p,'\0', sizeof(struct sock_fprog));
 
 	memset(&p, 0, sizeof(p));
 
diff -urNp -x '*.orig' dhcp-4.3.5.org/common/packet.c dhcp-4.3.5/common/packet.c
--- dhcp-4.3.5.org/common/packet.c	2016-09-27 21:16:50.000000000 +0200
+++ dhcp-4.3.5/common/packet.c	2021-04-01 22:09:12.139265409 +0200
@@ -145,6 +145,7 @@ void assemble_udp_ip_header (interface,
 
 	memset (&ip, 0, sizeof ip);
 
+	memset( &ip, '\0', sizeof ip);
 	/* Fill out the IP header */
 	IP_V_SET (&ip, 4);
 	IP_HL_SET (&ip, 20);
