diff -urNp -x '*.orig' dhcp-4.3.5.org/common/dispatch.c dhcp-4.3.5/common/dispatch.c
--- dhcp-4.3.5.org/common/dispatch.c	2016-09-27 21:16:50.000000000 +0200
+++ dhcp-4.3.5/common/dispatch.c	2021-04-01 22:09:11.449254169 +0200
@@ -410,7 +410,6 @@ void cancel_timeout (where, what)
 	}
 }
 
-#if defined (DEBUG_MEMORY_LEAKAGE_ON_EXIT)
 void cancel_all_timeouts ()
 {
 	struct timeout *t, *n;
@@ -432,4 +431,3 @@ void relinquish_timeouts ()
 		dfree(t, MDL);
 	}
 }
-#endif
