--- dhcp-3.1.0a3/client/scripts/linux.orig	2007-04-16 17:29:17.000000000 +0200
+++ dhcp-3.1.0a3/client/scripts/linux	2007-04-16 18:04:52.000000000 +0200
@@ -65,6 +65,20 @@
 	exit $exit_status
 }
 
+if [ -n "$dhc_dbus" -a -x /usr/bin/dbus-send ]; then
+	/usr/bin/dbus-send \
+		--system \
+		--dest=com.redhat.dhcp \
+		--type=method_call \
+		/com/redhat/dhcp/$interface \
+		com.redhat.dhcp.set \
+		'string:'"$(unset PATH SHLVL _ PWD dhc_dbus ; env)"
+
+	if [ $(($dhc_dbus & 31)) -eq 31 ]; then
+		exit 0
+	fi
+fi
+
 # Invoke the local dhcp client enter hooks, if they exist.
 if [ -f /etc/dhclient-enter-hooks ]; then
 	exit_status=0
