--- dhcp-3.0.3/client/scripts/linux.orig	2006-02-08 23:30:26.277493500 +0100
+++ dhcp-3.0.3/client/scripts/linux	2006-02-08 23:32:49.746459750 +0100
@@ -22,6 +22,22 @@
 # 4. TIMEOUT not tested. ping has a flag I don't know, and I'm suspicious
 # of the $1 in its args.
 
+if [ -n "${dhc_dbus}" ]; then
+  if [ -x /usr/bin/dbus-send ]; then
+    /usr/bin/dbus-send \
+      --system \
+      --dest=com.redhat.dhcp \
+      --type=method_call \
+      /com/redhat/dhcp/$interface \
+      com.redhat.dhcp.set \
+      'string:'"`env | /bin/egrep -v '^(PATH|SHLVL|_|PWD|dhc_dbus)\='`";
+
+    if (( ( dhc_dbus & 31 ) == 31 )); then
+      exit 0;
+    fi;
+  fi;
+fi;
+
 make_resolv_conf() {
   if [ "x$new_domain_name" != x ] && [ x"$new_domain_name_servers" != x ]; then
     echo search $new_domain_name >/etc/resolv.conf
