diff -up dhcp-4.0.0/client/dhclient.c.fast dhcp-4.0.0/client/dhclient.c
--- dhcp-4.0.0/client/dhclient.c.fast	2007-12-29 06:07:14.000000000 -1000
+++ dhcp-4.0.0/client/dhclient.c	2007-12-29 06:08:07.000000000 -1000
@@ -3624,10 +3624,7 @@ isc_result_t dhclient_interface_startup_
 		ip -> flags |= INTERFACE_RUNNING;
 		for (client = ip -> client; client; client = client -> next) {
 			client -> state = S_INIT;
-			/* Set up a timeout to start the initialization
-			   process. */
-			add_timeout (cur_time + random () % 5,
-				     state_reboot, client, 0, 0);
+			add_timeout (cur_time, state_reboot, client, 0, 0);
 		}
 	}
 	return ISC_R_SUCCESS;
