--- dhcp-3.0.5/dst/dst_api.c.prototypes	2007-04-01 16:29:10.000000000 -0400
+++ dhcp-3.0.5/dst/dst_api.c	2007-04-01 16:29:10.000000000 -0400
@@ -58,6 +58,10 @@
 
 #include "dst_internal.h"
 
+/* prototypes */
+extern int b64_pton(char const *src, u_char *target, size_t targsize);
+extern int b64_ntop(u_char const *src, size_t srclength, char *target, size_t targsize);
+
 /* static variables */
 static int done_init = 0;
 dst_func *dst_t_func[DST_MAX_ALGS];
--- dhcp-3.0.5/dst/dst_support.c.prototypes	2001-02-22 02:22:08.000000000 -0500
+++ dhcp-3.0.5/dst/dst_support.c	2007-04-01 16:29:10.000000000 -0400
@@ -32,6 +32,10 @@
 
 #include "dst_internal.h"
 
+/* prototypes */
+extern int b64_pton(char const *src, u_char *target, size_t targsize);
+extern int b64_ntop(u_char const *src, size_t srclength, char *target, size_t targsize);
+
 /*
  * dst_s_conv_bignum_u8_to_b64
  *	This function converts binary data stored as a u_char[] to a
--- dhcp-3.0.5/includes/minires/minires.h.prototypes	2004-06-10 13:59:37.000000000 -0400
+++ dhcp-3.0.5/includes/minires/minires.h	2007-04-01 16:29:10.000000000 -0400
@@ -65,7 +65,6 @@
 #define res_protocolname MRres_protocolname
 #define res_servicename MRres_servicename
 #define ns_datetosecs MRns_datetosecs
-#define b64_pton MRb64_pton
 #define res_ninit minires_ninit
 #define res_randomid MRres_randomid
 #define res_findzonecut MRres_findzonecut
--- dhcp-3.0.5/minires/res_mkupdate.c.prototypes	2007-04-01 16:29:10.000000000 -0400
+++ dhcp-3.0.5/minires/res_mkupdate.c	2007-04-01 16:30:12.000000000 -0400
@@ -49,6 +49,10 @@
 #include "minires/minires.h"
 #include "arpa/nameser.h"
 
+/* prototypes */
+extern int b64_pton(char const *src, u_char *target, size_t targsize);
+int dn_comp(const char *src, u_char *dst, unsigned dstsiz, u_char **dnptrs, u_char **lastdnptr);
+
 /* Options.  Leave them on. */
 #define DEBUG
 #define MAXPORT 1024
--- dhcp-3.0.5/minires/res_comp.c.prototypes	2004-06-10 13:59:42.000000000 -0400
+++ dhcp-3.0.5/minires/res_comp.c	2007-04-01 16:31:02.000000000 -0400
@@ -91,6 +91,11 @@
 #include "minires/minires.h"
 #include "arpa/nameser.h"
 
+/* prototypes */
+int ns_name_uncompress(const u_char *msg, const u_char *eom, const u_char *src, char *dst, size_t dstsiz);
+int ns_name_compress(const char *src, u_char *dst, size_t dstsiz, const u_char **dnptrs, const u_char **lastdnptr);
+int ns_name_skip(const u_char **ptrptr, const u_char *eom);
+
 /*
  * Expand compressed domain name 'comp_dn' to full domain name.
  * 'msg' is a pointer to the begining of the message,
--- dhcp-3.0.5/minires/res_init.c.prototypes	2004-06-10 13:59:43.000000000 -0400
+++ dhcp-3.0.5/minires/res_init.c	2007-04-01 16:31:38.000000000 -0400
@@ -102,6 +102,7 @@
 #define DEBUG
 
 static void res_setoptions (res_state, const char *, const char *);
+u_int res_randomid(void);
 
 #ifdef RESOLVSORT
 static const char sort_mask[] = "/&";
