--- dhcp-3.0.5/server/confpars.c.failover	2007-03-30 16:28:08.000000000 -0400
+++ dhcp-3.0.5/server/confpars.c	2007-03-30 16:32:39.000000000 -0400
@@ -998,10 +998,17 @@
 		parse_warn (cfile, "peer address may not be omitted");
 
 	/* XXX - when/if we get a port number assigned, just set as default */
+	/* See Red Hat Bugzilla 167292:
+	 * we do now:  dhcp-failover 647/tcp
+	 *             dhcp-failover 647/udp
+	 *             dhcp-failover 847/tcp
+	 *             dhcp-failover 847/udp
+	 * IANA registration by Bernard Volz <volz@cisco.com>
+	 */
 	if (!peer -> me.port)
-		parse_warn (cfile, "local port may not be omitted");
+		peer -> me.port = 647;
 	if (!peer -> partner.port)
-		parse_warn (cfile, "peer port may not be omitted");
+		peer -> partner.port = 847;
 
 	if (peer -> i_am == primary) {
 	    if (!peer -> hba) {
