diff -up dhcp-4.0.0/client/dhclient.c.usage dhcp-4.0.0/client/dhclient.c
--- dhcp-4.0.0/client/dhclient.c.usage	2008-02-13 16:20:30.000000000 -1000
+++ dhcp-4.0.0/client/dhclient.c	2008-02-13 16:21:05.000000000 -1000
@@ -882,21 +882,30 @@ main(int argc, char **argv) {
 
 static void usage ()
 {
-	log_info ("%s %s", message, PACKAGE_VERSION);
-	log_info (copyright);
-	log_info (arr);
-	log_info (url);
+	printf ("%s %s\n", message, PACKAGE_VERSION);
+	printf (copyright);
+	printf ("\n");
+	printf (arr);
+	printf ("\n");
+	printf (url);
+	printf ("\n");
 
-	log_error ("Usage: dhclient %s %s",
+	printf ("Usage: dhclient %s %s",
 #ifdef DHCPv6
                    "[-4|-6] [-1dvrx] [-nw] [-p <port>]",
 #else /* DHCPv6 */
                    "[-1dvrx] [-nw] [-p <port>]",
 #endif /* DHCPv6 */
-		   "[-s server]");
-	log_error ("                [-cf config-file] [-lf lease-file]%s",
-		   "[-pf pid-file] [-e VAR=val]");
-	log_fatal ("                [-sf script-file] [interface]");
+		   "[-s server]\n");
+	printf ("                [-cf config-file] [-lf lease-file]%s",
+		   "[-pf pid-file] [-e VAR=val]\n");
+	printf ("                [ -I <dhcp-client-identifier> ] [-B]\n");
+	printf ("                [ -H <host-name> | -F <fqdn.fqdn> ] [ -T <timeout> ]\n");
+	printf ("                [ -V <vendor-class-identifier> ]\n");
+	printf ("                [ -R <request option list> ]\n");
+	printf ("                [-sf script-file] [interface]\n");
+
+	exit (EXIT_FAILURE);
 }
 
 isc_result_t find_class (struct class **c,
