--- dhcp-4.0.0//client/clparse.c.nmcrash	2008-04-01 10:25:39.000000000 -1000
+++ dhcp-4.0.0//client/clparse.c	2008-04-01 10:26:20.000000000 -1000
@@ -227,6 +227,9 @@ int read_client_conf_file (const char *n
 	cfile = (struct parse *)0;
 	new_parse (&cfile, file, (char *)0, 0, path_dhclient_conf, 0);
 
+	if (!cfile)
+		return ISC_R_BADPARSE;
+
 	do {
 		token = peek_token (&val, (unsigned *)0, cfile);
 		if (token == END_OF_FILE)
